tests/test_mr.py

```diff
@@ -161,35 +161,55 @@ def test_reg_mw_type1(self):
         """
         for ctx, attr, attr_ex in self.devices:
             with PD(ctx) as pd:
-                with MW(pd, e.IBV_MW_TYPE_1):
-                    pass
+                try:
+                    with MW(pd, e.IBV_MW_TYPE_1):
+                        pass
+                except PyverbsRDMAError as ex:
+                    if 'not implemented' in str(ex):
+                        raise unittest.SkipTest('Create memory window of type 1 is not supported')
+                    raise ex
 
     def test_reg_mw_type2(self):
         """
         Test ibv_alloc_mw() for type 2 MW
         """
         for ctx, attr, attr_ex in self.devices:
             with PD(ctx) as pd:
-                with MW(pd, e.IBV_MW_TYPE_2):
-                    pass
+                try:
+                    with MW(pd, e.IBV_MW_TYPE_2):
+                        pass
+                except PyverbsRDMAError as ex:
+                    if 'not implemented' in str(ex):
+                        raise unittest.SkipTest('Create memory window of type 2 is not supported')
+                    raise ex
 
     def test_dereg_mw_type1(self):
         """
         Test ibv_dealloc_mw() for type 1 MW
         """
         for ctx, attr, attr_ex in self.devices:
             with PD(ctx) as pd:
-                with MW(pd, e.IBV_MW_TYPE_1) as mw:
-                    mw.close()
+                try:
+                    with MW(pd, e.IBV_MW_TYPE_1) as mw:
+                        mw.close()
+                except PyverbsRDMAError as ex:
```

Comment:
So if MW dereg failed, you supress pyverbs' exception and not raising anything instead.

Reply:
No, I'm raising the 'ex', please see line 198.

Reply:
This should directly test for EOPNOTSUPP, not via a string compare.

Reply:
> This should directly test for EOPNOTSUPP, not via a string compare.

Well, the error in this case is `#define ENOSYS 88	/* Function not implemented */` - Please see below:

```
Traceback (most recent call last):
  File "/home/rdma-core/tests/test_mr.py", line 182, in test_dereg_mw_type1
    with MW(pd, e.IBV_MW_TYPE_1) as mw:
  File "mr.pyx", line 124, in pyverbs.mr.MW.__init__
pyverbs.pyverbs_error.PyverbsRDMAError: Failed to allocate MW. Errno: 38, Function not implemented
```

currently, the only way to test is via comparing the string from the exception, unless you meat to change the PyverbsError to store the errno value - so later in the code, I can do the following:

```
try:
    with MW(pd, e.IBV_MW_TYPE_1):
        pass
    except PyverbsRDMAError as ex:
        if ex.error_code == errno.ENOSYS:
            raise unittest.SkipTest('Create memory window of type 1 is not supported')
        raise ex
```

Reply:
The error code is supposed to EOPNOTSUPP, returning ENOSYS is a bug in libibverbs or the provider that we should fix.

Reply:
And yes to changing to store the error code directly in the PyverbsRDMAError

Reply:
We already have a PyverbsRDMAErrno which returns PyverbsRDMAError with the errno in its message. So maybe it's better to re-define it as a class, inheriting from PyverbsRDMAError and storing the errno. (must check what is the consequences of this change - if any)

Reply:
> We already have a PyverbsRDMAErrno which returns PyverbsRDMAError with the errno in its message. So maybe it's better to re-define it as a class, inheriting from PyverbsRDMAError and storing the errno. (must check what is the consequences of this change - if any)

Based on [1] reference Cython classes can not inherit from normal classes, which probably describes why PyverbsRDMAErrno is not defined as a class.

`Normal Python classes can inherit from cdef classes, but not the other way around. Cython requires to know the complete inheritance hierarchy in order to lay out their C structs, and restricts it to single inheritance. Normal Python classes, on the other hand, can inherit from any number of Python classes and extension types, both in Cython code and pure Python code.
`

I plan to store the errno value to be part of the PyverbsError base class, I think this change makes more sense, first because the "error_code" is used in the base class to build the msg string, second maybe in the future we want to use the errno code with PyverbsUserError?

@EdwardSro : Thoughts?

[1] - https://cython.readthedocs.io/en/latest/src/tutorial/cdef_classes.html

Reply:
Oh right. Then yes, I would go in that way.
Regarding PyverbsUserError, it should be used for errors made solely by pyverbs / pyverbs' users (e.g. passing wrong device name / an exception for feature that's partially supported **in pyverbs** etc.), so errno shouldn't be relevant for PyverbsUserError (if it is, then most likely PyverbsRDMAError or other exception should be used)

Reply:
While this is OK, it seems a little off, and maybe a followup should revise how we handle this.

But basically, tests should only be skipped from decorators that delcare some required test capability. ie running the  test_reg_mw_wrong_type tests without MW support is just as wrong.

Probing the capability from a EOPNOTSUPP in a decorator requires some more fancy scheme, and requires the tests be moved to the new scheme where the test framework so each test runs on a single ibv_context and the decorator can then check its support.